x/service/keeper/msg_server_add_service_test.go

```diff
@@ -0,0 +1,197 @@
+		},
+		{
+			desc:    "invalid - empty service ID",
+			setup:   func(t *testing.T) {},
+			address: sample.AccAddress(),
+			service: sharedtypes.Service{
+				Id:   "", // explicitly set to empty string
+				Name: "service 1",
+			},
+			expectedError: types.ErrServiceMissingID,
+		},
+		{
+			desc:    "invalid - missing service name",
+			setup:   func(t *testing.T) {},
+			address: sample.AccAddress(),
+			service: sharedtypes.Service{
+				Id: "svc1",
+				// Explicitly omitting Name field
+			},
+			expectedError: types.ErrServiceMissingName,
+		},
+		{
+			desc:    "invalid - empty service name",
+			setup:   func(t *testing.T) {},
+			address: sample.AccAddress(),
+			service: sharedtypes.Service{
+				Id:   "svc1",
+				Name: "", // explicitly set to empty string
+			},
+			expectedError: types.ErrServiceMissingName,
+		},
+		{
+			desc:          "invalid - service already exists (same service supplier)",
+			setup:         func(t *testing.T) {},
+			address:       addr,
+			service:       preExistingService,
+			expectedError: types.ErrServiceAlreadyExists,
+		},
+		{
+			desc:          "invalid - service already exists (different service supplier)",
+			setup:         func(t *testing.T) {},
+			address:       sample.AccAddress(),
+			service:       preExistingService,
+			expectedError: types.ErrServiceAlreadyExists,
+		},
+		{
+			desc:          "invalid - no spendable coins",
+			setup:         func(t *testing.T) {},
+			address:       sample.AccAddress(),
+			service:       svc1,
+			expectedError: types.ErrServiceNotEnoughFunds,
+		},
+		{
+			desc: "invalid - insufficient upokt balance",
+			setup: func(t *testing.T) {
+				// Add 999999999 upokt to the account (one less than AddServiceFee)
+				keepertest.AddAccToAccMapCoins(t, validAddr2, "upokt", oneUPOKTGreaterThanFee-2)
+			},
+			address:       validAddr2,
+			service:       svc1,
+			expectedError: types.ErrServiceNotEnoughFunds,
+		},
+		{
+			desc: "invalid - account has exactly AddServiceFee",
+			setup: func(t *testing.T) {
+				// Add the exact fee in upokt to the account
+				keepertest.AddAccToAccMapCoins(t, validAddr2, "upokt", types.DefaultAddServiceFee)
+			},
+			address:       validAddr2,
+			service:       svc1,
+			expectedError: types.ErrServiceNotEnoughFunds,
+		},
+		{
+			desc: "invalid - sufficient balance of different denom",
+			setup: func(t *testing.T) {
+				// Adds 10000000001 wrong coins to the account
+				keepertest.AddAccToAccMapCoins(t, validAddr2, "wrong", oneUPOKTGreaterThanFee)
+			},
+			address:       validAddr2,
+			service:       svc1,
+			expectedError: types.ErrServiceNotEnoughFunds,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			tt.setup(t)
+			_, err := srv.AddService(ctx, &types.MsgAddService{
+				Address: tt.address,
+				Service: tt.service,
+			})
+			if tt.expectedError != nil {
+				// Using ErrorAs as wrapping the error sometimes gives errors with ErrorIs
```

Comment:
```suggestion
				// TODO_IMPROVE: more consistent and robust error assertion.
				// Using `ErrorAs()` because using `ErrorIs()` sometimes causes issues.
```